### PR TITLE
stm32/isr: Add option to run flash & uart isr function from ram.

### LINKED ISSUE #8386: mpremote cannot cp files to stm32wb55 nucleo on stlink/com port.
When connecting to the nucleo stm32wb55 repl on the stlink (usb <-> uart) port file transfers with mpremote above a certain (small) size always fail. 

In this case, transferring a ~23K file hangs for about 15 seconds then fails with the below error.
```
micropython\ports\stm32\boards\NUCLEO_WB55>py ..\..\..\..\tools\mpremote\mpremote.py cp rfcore_firmware.py :
cp rfcore_firmware.py :rfcore_firmware.py
Traceback (most recent call last):
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote/../..\pyboard.py", line 582, in filesystem_command
    op(src, dest2)
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote/../..\pyboard.py", line 522, in fs_put
    self.exec_("w(" + repr(data) + ")")
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote/../..\pyboard.py", line 465, in exec_
    ret, ret_err = self.exec_raw(command, data_consumer=data_consumer)
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote/../..\pyboard.py", line 456, in exec_raw
    self.exec_raw_no_follow(command)
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote/../..\pyboard.py", line 434, in exec_raw_no_follow
    return self.raw_paste_write(command_bytes)
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote/../..\pyboard.py", line 412, in raw_paste_write
    raise PyboardError("could not complete raw paste: {}".format(data))
pyboard.PyboardError: could not complete raw paste: b''

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote.py", line 6, in <module>
    sys.exit(main.main())
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote\main.py", line 520, in main
    do_filesystem(pyb, args)
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote\main.py", line 298, in do_filesystem
    pyboard.filesystem_command(pyb, args)
  File "W:\home\anl\tempt\src\micropython\tools\mpremote\mpremote/../..\pyboard.py", line 598, in filesystem_command
    print(str(er.args[2], "ascii"))
IndexError: tuple index out of range
```

It fails in the same way after a similar timeout every time.

The same file transfer works perfectly on the direct usb port.

Transferring a 259 byte board.json works. 
Transferring a 575 byte mpconfigboard.mk fails.

Note this is on Windows so may be related to https://github.com/micropython/micropython/issues/8358 - I will test shortly if this PR resolves (hides) the issue.

### PR DESCRIPTION
**Pull request by andrewleech**
This allows uart rx to function while flash erase/writes operations
are under way, preventing lost data so long as it fits in the uart
rx buffer.

This is a more general implementation of https://github.com/micropython/micropython/pull/8713 which is intended to fix https://github.com/micropython/micropython/issues/8386

Enable in mpconfigboard.mk:
MICROPY_USE_RAM_ISR_UART_FLASH_FN = 1

It's been enabled on the NUCLEO_WB55 but should equally apply to any board with its default repl on a uart port without flow control (most nucleo boards with onboard stlink).

This PR is currently sitting on top of / includes https://github.com/micropython/micropython/pull/9053 which makes the diff somewhat messy to view on github unfortunately.

It also includes a commit to pre-process linker scripts.
This allows the use of command line -D defines and/or including header files to switch features on/off in the linker scripts. 

It's used in the `ld` files to select between `common_isr.ld` and `common_isr_ram.ld` snippet.

### CODE REVIEW

**Comment by dlech** [comment]
Doesn't this require that *all* ISR functions and their depedencies be moved to RAM? Or does this somehow rely on interrupt priorities or something like that to ensure that only certain interrupts are enabled during flashing. For example, what happens if someone implements a GPIO interrupt handler in Python and that is triggered while flashing?

**Comment by andrewleech** [comment]
@dlech yes it's a good question that I'm still not 100% certain about. The irq priorities might be enough protection, that's why pend and systick are in the list being higher prio. 

It would be good to test with a high frequency timer interrupt I guess to see if that breaks it. If that does lock up we could move the actual flash operation inside an irq at the appropriate priority, that would then definitely block lower prio irqs.

**Comment by andrewleech** [comment]
So the pre-processing of linker scripts comes with an undesirable side-effect - lines like `*lib/mbedtls/*(.text* .rodata*)` throw errors because it include `/*` in the middle of the line. While it can be worked around with `*lib/mbedtls/?*(.text* .rodata*)` that's a bit difficult to communicate / document.

As such I've reworked this to PR to no longer use gcc pre-processing of linker files.

FWIW if anyone's interested in the linker preproc it can still be viewed here: https://github.com/micropython/micropython/commit/ffc88fc487ed1a871406582e35a58b5f743a9032

**Comment by iabdalkader** [comment]
Hi, should this be enabled by default for every STM board, including ones that don't even use the internal flash storage for the filesystem ?

**Comment by andrewleech** [comment]
The latest commit (0abc1d9) has been tested on the USB/UART interface of NUCLEO_WB55 to copy a ~ 66KB file with mpremote. Previously this would fail every time.

**Comment by andrewleech** [comment]
> Hi, should this be enabled by default for every STM board, including ones that don't even use the internal flash storage for the filesystem ?

At the moment I believe the only "significant" value of this change is allowing uart repl to work while flash is busy. 

I did question whether having some ISR's in ram would increase performance for all stm's but have heard that the internal flash cache (ART) probably already provides a similar level of performance gain automatically.

**Comment by andrewleech** [comment]
TODO re-test whether flash functions need to be in ram; this costs ram in use that may not actually be needed.

**Comment by andrewleech** [comment]
Updated and rebased to current master. 

As per https://github.com/micropython/micropython/commit/091aa93ab03a48b8d6947db490691c8863505414#r143545503 I've cleaned up the mechanism for including the correct linker snippet. Re-tested on WB55 to confirm it still works correctly.

I checked through the boards I'd modified and found quite a few that aren't using the internal filesystem, so they've been reverted.

I then rechecked all boards for any new ones that needed it added, updating them as needed.
This following was used for search for all boards which:
* declare `UART_REPL` in mpconfigboard.h
* and do not declare `MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE (0)`
* and do not alreadyt have `MICROPY_USE_RAM_ISR_UART_FLASH_FN ` in mpconfigboard.mk
```
for h in $(grep -l -R UART_REPL ports/stm32/boards/*); do 
grep -q "MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE.*\(0\)" "$h" || \
grep --files-without-match MICROPY_USE_RAM_ISR_UART_FLASH_FN "${h%.*}.mk";
done
```

As per:
> TODO re-test whether flash functions need to be in ram; this costs ram in use that may not actually be needed.

I tested a build removing all the FLASH related from the ram linker listing, found the problem came straight back - copying file causes lockups. Bringing the FLASH functions back into the ram/isr linker fixes it again.

**Comment by andrewleech** [comment]
Currently I've removed the FLASH_ISR MEMORY section in linker scripts as the block is now ALIASED into flash or ram as appropriate.
On many (older) boards though, this memory section needs to be limited to 16K as the filesystem is directly after it (to meet sector size needs). 
So without this 16k limited size, if the FLASH_ISR overflows in future it won't get caught by the linker and instead break the filesystem.

So FLASH_ISR (or similar) needs to be added back to the MEMORY section of linker files, at least for boards with a "middle" filesytem, then the ALIAS point to to that instead of just FLASH.

**Comment by andrewleech** [comment]
I've updated this to preserve the size-limited first sector in linker where used to ensure that any future overflow is detected correctly.

All boards were compiled to ensure they all _can_ still be compiled:
```
~/micropython $ for board in $(cd ports/stm32/boards; find * -type d); do [ -e ports/stm32/boards/$board/mpconfigboard.mk ] && make -C ports/stm32 BOARD="$board" -j18 BUILD=../../after/build-$board; done
```
and
```
~/micropython $ for board in $(cd ports/stm32/boards; find * -type d); do [ -e ports/stm32/boards/$board/mpconfigboard.mk ] && make -C ports/stm32 BOARD="$board" MBOOT=1 -j18 BUILD=../../after-mboot/build-$board; done
```

File: ports/stm32/Makefile [hunk 1]
```diff
@@ -143,7 +143,15 @@ CFLAGS += -DMICROPY_FLOAT_IMPL=MICROPY_FLOAT_IMPL_FLOAT
 CFLAGS += -fsingle-precision-constant
 endif
 
-LDFLAGS += -nostdlib -L $(LD_DIR) $(addprefix -T,$(LD_FILES)) -Wl,-Map=$(@:.elf=.map) -Wl,--cref
+# Configure linker include dir for ram/rom isr support
+ifeq ($(MICROPY_USE_RAM_ISR_UART_FLASH_FN),1)
```
Comment (dpgeorge) [inline ports/stm32/Makefile:147]: How about calling this option `MICROPY_HW_ENABLE_ISR_UART_FLASH_FUNCS_IN_RAM`?

File: ports/stm32/boards/STM32F769DISC/f769_qspi.ld [hunk 2]
```diff
@@ -38,8 +39,16 @@ _ram_end = ORIGIN(RAM) + LENGTH(RAM);
 _heap_start = _ebss; /* heap starts just after statically allocated memory */
 _heap_end = _sstack;
 
+/* Filesystem cache in RAM, and storage in flash */
+_micropy_hw_internal_flash_storage_ram_cache_start = ORIGIN(DTCM);
+_micropy_hw_internal_flash_storage_ram_cache_end = ORIGIN(DTCM) + LENGTH(DTCM);
+_micropy_hw_internal_flash_storage_start = ORIGIN(FLASH_FS);
+_micropy_hw_internal_flash_storage_end = ORIGIN(FLASH_FS) + LENGTH(FLASH_FS);
+
+
```
Comment (dpgeorge) [inline ports/stm32/boards/STM32F769DISC/f769_qspi.ld:48]: please only have a single blank line here

File: ports/stm32/boards/common_isr_ram/common_isr.ld [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+.isr_vector :
+{
+    . = ALIGN(4);
+    __isr_vector_ram_start = .;
+    KEEP(*(.isr_vector)) /* Startup code */
+
+    /* These functions need to run from ram to enable uart
+       reception during flash erase/write operations.
+       Defining them here ensures they're copied from
+       flash (in ram_irq.c) along with the isr_vector above.
```
Comment (dpgeorge) [inline ports/stm32/boards/common_isr_ram/common_isr.ld:13]: `ram_irq.c` no longer exists

File: ports/stm32/boards/common_isr_ram/common_isr.ld [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    *(.text.FLASH_WaitForLastOperation)
+    *(.text.HAL_FLASHEx_Erase)
+    *(.text.HAL_GetTick)
+
+    __isr_vector_ram_end = .;
+    . = ALIGN(4);
+
+} >RAM AT >FLASH_ISR
+
+_start_init_isr = LOADADDR(.isr_vector); /* Used by the start-up code to initialise data */
```
Comment (dpgeorge) [inline ports/stm32/boards/common_isr_ram/common_isr.ld:34]: looks like this variable is unused

File: ports/stm32/boards/common_isr_ram/common_isr.ld [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+
+/* Used by the start-up code to initialise data */
+__isr_vector_flash_addr = LOADADDR(.isr_vector);
+__isr_vector_ram_addr = ADDR(.isr_vector);
+__isr_section_len = SIZEOF(.isr_vector);
```
Comment (dpgeorge) [inline ports/stm32/boards/common_isr_ram/common_isr.ld:40]: looks like the latter two variables here are unused

File: ports/stm32/boards/common_isr_rom/common_isr_extratext.ld [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+.isr_extratext :
+{
+    . = ALIGN(4);
+    /* This first flash block is 16K and the isr vectors only take up
+        about 400 bytes. So we pull in a couple of object files to pad it
+        out.
+
+        NOTE: If you update the list of files contained in .isr_extratext,
+        then be sure to also update smhal/Makefile where it forcibly
```
Comment (dpgeorge) [inline ports/stm32/boards/common_isr_rom/common_isr_extratext.ld:13]: please replace "smhal" with "stm32"

File: ports/stm32/boards/common_isr_rom/common_isr_extratext.ld [hunk 7]
```diff
@@ -0,0 +1,24 @@
+/* This linker script fragment is intended to be included in SECTIONS when
+   common_ifs.ld is used with common_isr.ld
+. */
+
+.isr_extratext :
+{
+    . = ALIGN(4);
+    /* This first flash block is 16K and the isr vectors only take up
```
Comment (dpgeorge) [inline ports/stm32/boards/common_isr_rom/common_isr_extratext.ld:8]: suggest adding a blank line between ALIGN and the comment

File: ports/stm32/boards/common_isr_rom/common_isr_extratext.ld [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        about 400 bytes. So we pull in a couple of object files to pad it
+        out.
+
+        NOTE: If you update the list of files contained in .isr_extratext,
+        then be sure to also update smhal/Makefile where it forcibly
+        builds each of these files with -Os
+    */
+
```
Comment (dpgeorge) [inline ports/stm32/boards/common_isr_rom/common_isr_extratext.ld:16]: suggest removing this blank line

File: ports/stm32/boards/STM32F769DISC/f769_qspi.ld [hunk 9]
```diff
@@ -38,8 +39,16 @@ _ram_end = ORIGIN(RAM) + LENGTH(RAM);
...
 ENTRY(Reset_Handler)
 
+REGION_ALIAS("FLASH_ISR", FLASH_APP);
```
Comment (dpgeorge) [inline ports/stm32/boards/STM32F769DISC/f769_qspi.ld:51]: The addition of this line belongs in the first commit, not the second.

**Comment by dpgeorge** [comment]
I tested this on a NUCLEO_WB55 with UART REPL, and it works really well, mpremote cp to the device works.

**Comment by dpgeorge** [comment]
I verified that PYBV10 and PYBD_SF6 firmware do not change with this PR.

File: ports/stm32/boards/common_isr_ram/common_isr.ld (hunk 4 above)
Comment (andrewleech) [reply ports/stm32/boards/common_isr_ram/common_isr.ld:34]: Ah you're right, seems I had a few different attempts worth of variables defined in here still sorry!

File: ports/stm32/boards/common_isr_rom/common_isr_extratext.ld (hunk 8 above)
Comment (andrewleech) [reply ports/stm32/boards/common_isr_rom/common_isr_extratext.ld:16]: Thanks that does make it clearer.

**Comment by andrewleech** [comment]
I've updated as suggested thanks @dpgeorge

**Comment by dpgeorge** [comment]
Thanks for updating, now merged!

**Pull request merged**